src/core/build_target.go

```diff
@@ -1068,29 +1068,42 @@ func (target *BuildTarget) AddProvide(language string, label BuildLabel) {
 
 // ProvideFor returns the build label that we'd provide for the given target.
 func (target *BuildTarget) ProvideFor(other *BuildTarget) []BuildLabel {
+	if p := target.provideFor(other); len(p) > 0 {
+		return p
+	}
+	return []BuildLabel{target.Label}
+}
+
+// provideFor is like ProvideFor but returns an empty slice if there is a direct dependency.
+// It's a small optimisation to save allocating extra slices.
+func (target *BuildTarget) provideFor(other *BuildTarget) []BuildLabel {
 	target.mutex.RLock()
 	defer target.mutex.RUnlock()
-	ret := []BuildLabel{}
-	if target.Provides != nil && len(other.Requires) != 0 {
-		// Never do this if the other target has a data or tool dependency on us.
-		for _, data := range other.Data {
-			if label := data.Label(); label != nil && *label == target.Label {
-				return []BuildLabel{target.Label}
-			}
-		}
-		if other.IsTool(target.Label) {
-			return []BuildLabel{target.Label}
+	if target.Provides == nil || len(other.Requires) == 0 {
+		return nil
+	}
+	// Never do this if the other target has a data or tool dependency on us.
+	for _, data := range other.Data {
+		if label := data.Label(); label != nil && *label == target.Label {
+			return nil
 		}
-		for _, require := range other.Requires {
-			if label, present := target.Provides[require]; present {
+	}
+	if other.IsTool(target.Label) {
+		return nil
+	}
+	var backingStorage [2]BuildLabel
+	var ret []BuildLabel
+	for _, require := range other.Requires {
+		if label, present := target.Provides[require]; present {
+			if ret == nil {
+				backingStorage[0] = label
+				ret = backingStorage[:1]
+			} else {
```

Comment:
I reckon this is the block where the optimisation is coming from?
- Why does `backingStorage` need a length of 2 if it's being sliced up to 1 here?
- Given the above, wouldn't `ret = append(ret,  label)` just do something similar the first time (allocating one item at first since it is empty).

I'm likely missing something here.